python/pyarrow/_flight.pyx

```diff
@@ -2516,6 +2517,32 @@ cdef class _ServerMiddlewareWrapper(ServerMiddleware):
             instance.call_completed(exception)
 
 
+cdef class _FlightServerFinalizer(_Weakrefable):
+    """
+    A finalizer that shuts down the server on destruction.
+
+    See ARROW-16597. If the server is still active at interpreter
+    exit, the process may segfault.
+    """
+
+    cdef:
+        shared_ptr[PyFlightServer] server
+
+    def finalize(self):
+        cdef:
+            PyFlightServer* server = self.server.get()
+            CStatus status
+        if server == NULL:
+            return
+        with nogil:
+            status = server.Shutdown()
+        check_flight_status(status)
+        with nogil:
+            status = server.Wait()
+        check_flight_status(status)
+        self.server.reset()
```

Comment:
`reset()` won't be called if there was an error above, could this be a problem?

Reply:
It shouldn't be a big deal, but I wrapped it in a try-except.